File: algorithms/blind75/1/hb_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        ans = True
        
        def hb_check(node):
            nonlocal ans
            if node == None: return 0
            l = hb_check(node.left)
            r = hb_check(node.right)
            if abs(l - r) > 1:
                ans = False
                return 0

            return 1 + max(l, r)

        hb_check(root)
        return ans

File: algorithms/blind75/1/hb_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool ans = true;
        checkhb(root, ans);
        return ans;
    }
    int checkhb(TreeNode* node, bool& ans) {
        if (node == nullptr) return 0;
        int l = checkhb(node->left, ans);
        int r = checkhb(node->right, ans);
        if (abs(l-r) > 1) {
            ans = false;
            return 0;
        } else {
            return 1 + max(l, r);
        }
    }
};
